# Multiple Regression

Simple Linear Regression:

$$y = \beta_0 + \beta_1X$$

Multiple Linear Regression:

$$y = \beta_0 + \beta_1X_1 + \beta_2X_2 + ...$$

Well studied field in statistics

Focus will be on what is relevant for Data Science - practical and relevant for prediction

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns

In [2]:
from sklearn.datasets import load_boston

In [3]:
boston_data = load_boston()

In [4]:
df = pd.DataFrame(boston_data.data, columns=boston_data.feature_names)

In [5]:
df.head()

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0  0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.0900  1.0  296.0   
1  0.02731   0.0   7.07   0.0  0.469  6.421  78.9  4.9671  2.0  242.0   
2  0.02729   0.0   7.07   0.0  0.469  7.185  61.1  4.9671  2.0  242.0   
3  0.03237   0.0   2.18   0.0  0.458  6.998  45.8  6.0622  3.0  222.0   
4  0.06905   0.0   2.18   0.0  0.458  7.147  54.2  6.0622  3.0  222.0   

   PTRATIO       B  LSTAT  
0     15.3  396.90   4.98  
1     17.8  396.90   9.14  
2     17.8  392.83   4.03  
3     18.7  394.63   2.94  
4     18.7  396.90   5.33

In [6]:
df.shape

(506, 13)

In [7]:
X = df

In [8]:
y = boston_data.target

In [9]:
y

array([ 24. ,  21.6,  34.7,  33.4,  36.2,  28.7,  22.9,  27.1,  16.5,
        18.9,  15. ,  18.9,  21.7,  20.4,  18.2,  19.9,  23.1,  17.5,
        20.2,  18.2,  13.6,  19.6,  15.2,  14.5,  15.6,  13.9,  16.6,
        14.8,  18.4,  21. ,  12.7,  14.5,  13.2,  13.1,  13.5,  18.9,
        20. ,  21. ,  24.7,  30.8,  34.9,  26.6,  25.3,  24.7,  21.2,
        19.3,  20. ,  16.6,  14.4,  19.4,  19.7,  20.5,  25. ,  23.4,
        18.9,  35.4,  24.7,  31.6,  23.3,  19.6,  18.7,  16. ,  22.2,
        25. ,  33. ,  23.5,  19.4,  22. ,  17.4,  20.9,  24.2,  21.7,
        22.8,  23.4,  24.1,  21.4,  20. ,  20.8,  21.2,  20.3,  28. ,
        23.9,  24.8,  22.9,  23.9,  26.6,  22.5,  22.2,  23.6,  28.7,
        22.6,  22. ,  22.9,  25. ,  20.6,  28.4,  21.4,  38.7,  43.8,
        33.2,  27.5,  26.5,  18.6,  19.3,  20.1,  19.5,  19.5,  20.4,
        19.8,  19.4,  21.7,  22.8,  18.8,  18.7,  18.5,  18.3,  21.2,
        19.2,  20.4,  19.3,  22. ,  20.3,  20.5,  17.3,  18.8,  21.4,
        15.7,  16.2,

# Statsmodels

In [11]:
import statsmodels.api as sm
import statsmodels.formula.api as smf

We need to add a constant term to allow statsmodel.api to calculate the bias / intercepts.

In [12]:
X_constant = sm.add_constant(X)

In [13]:
pd.DataFrame(X_constant)

const      CRIM    ZN  INDUS  CHAS    NOX     RM    AGE     DIS   RAD  \
0      1.0   0.00632  18.0   2.31   0.0  0.538  6.575   65.2  4.0900   1.0   
1      1.0   0.02731   0.0   7.07   0.0  0.469  6.421   78.9  4.9671   2.0   
2      1.0   0.02729   0.0   7.07   0.0  0.469  7.185   61.1  4.9671   2.0   
3      1.0   0.03237   0.0   2.18   0.0  0.458  6.998   45.8  6.0622   3.0   
4      1.0   0.06905   0.0   2.18   0.0  0.458  7.147   54.2  6.0622   3.0   
5      1.0   0.02985   0.0   2.18   0.0  0.458  6.430   58.7  6.0622   3.0   
6      1.0   0.08829  12.5   7.87   0.0  0.524  6.012   66.6  5.5605   5.0   
7      1.0   0.14455  12.5   7.87   0.0  0.524  6.172   96.1  5.9505   5.0   
8      1.0   0.21124  12.5   7.87   0.0  0.524  5.631  100.0  6.0821   5.0   
9      1.0   0.17004  12.5   7.87   0.0  0.524  6.004   85.9  6.5921   5.0   
10     1.0   0.22489  12.5   7.87   0.0  0.524  6.377   94.3  6.3467   5.0   
11     1.0   0.11747  12.5   7.87   0.0  0.524  6.009   82.9  6.2267   5.0   
12     1.0   0.09378  12.5   7.87   0.0  0.524  5.889   39.0  5.4509   5.0   
13     1.0   0.62976   0.0   8.14   0.0  0.538  5.949   61.8  4.7075   4.0   
14     1.0   0.63796   0.0   8.14   0.0  0.538  6.096   84.5  4.4619   4.0   
15     1.0   0.62739   0.0   8.14   0.0  0.538  5.834   56.5  4.4986   4.0   
16     1.0   1.05393   0.0   8.14   0.0  0.538  5.935   29.3  4.4986   4.0   
17     1.0   0.78420   0.0   8.14   0.0  0.538  5.990   81.7  4.2579   4.0   
18     1.0   0.80271   0.0   8.14   0.0  0.538  5.456   36.6  3.7965   4.0   
19     1.0   0.72580   0.0   8.14   0.0  0.538  5.727   69.5  3.7965   4.0   
20     1.0   1.25179   0.0   8.14   0.0  0.538  5.570   98.1  3.7979   4.0   
21     1.0   0.85204   0.0   8.14   0.0  0.538  5.965   89.2  4.0123   4.0   
22     1.0   1.23247   0.0   8.14   0.0  0.538  6.142   91.7  3.9769   4.0   
23     1.0   0.98843   0.0   8.14   0.0  0.538  5.813  100.0  4.0952   4.0   
24     1.0   0.75026   0.0   8.14   0.0  0.538  5.924   94.1  4.3996   4.0   
25     1.0   0.84054   0.0   8.14   0.0  0.538  5.599   85.7  4.4546   4.0   
26     1.0   0.67191   0.0   8.14   0.0  0.538  5.813   90.3  4.6820   4.0   
27     1.0   0.95577   0.0   8.14   0.0  0.538  6.047   88.8  4.4534   4.0   
28     1.0   0.77299   0.0   8.14   0.0  0.538  6.495   94.4  4.4547   4.0   
29     1.0   1.00245   0.0   8.14   0.0  0.538  6.674   87.3  4.2390   4.0   
..     ...       ...   ...    ...   ...    ...    ...    ...     ...   ...   
476    1.0   4.87141   0.0  18.10   0.0  0.614  6.484   93.6  2.3053  24.0   
477    1.0  15.02340   0.0  18.10   0.0  0.614  5.304   97.3  2.1007  24.0   
478    1.0  10.23300   0.0  18.10   0.0  0.614  6.185   96.7  2.1705  24.0   
479    1.0  14.33370   0.0  18.10   0.0  0.614  6.229   88.0  1.9512  24.0   
480    1.0   5.82401   0.0  18.10   0.0  0.532  6.242   64.7  3.4242  24.0   
481    1.0   5.70818   0.0  18.10   0.0  0.532  6.750   74.9  3.3317  24.0   
482    1.0   5.73116   0.0  18.10   0.0  0.532  7.061   77.0  3.4106  24.0   
483    1.0   2.81838   0.0  18.10   0.0  0.532  5.762   40.3  4.0983  24.0   
484    1.0   2.37857   0.0  18.10   0.0  0.583  5.871   41.9  3.7240  24.0   
485    1.0   3.67367   0.0  18.10   0.0  0.583  6.312   51.9  3.9917  24.0   
486    1.0   5.69175   0.0  18.10   0.0  0.583  6.114   79.8  3.5459  24.0   
487    1.0   4.83567   0.0  18.10   0.0  0.583  5.905   53.2  3.1523  24.0   
488    1.0   0.15086   0.0  27.74   0.0  0.609  5.454   92.7  1.8209   4.0   
489    1.0   0.18337   0.0  27.74   0.0  0.609  5.414   98.3  1.7554   4.0   
490    1.0   0.20746   0.0  27.74   0.0  0.609  5.093   98.0  1.8226   4.0   
491    1.0   0.10574   0.0  27.74   0.0  0.609  5.983   98.8  1.8681   4.0   
492    1.0   0.11132   0.0  27.74   0.0  0.609  5.983   83.5  2.1099   4.0   
493    1.0   0.17331   0.0   9.69   0.0  0.585  5.707   54.0  2.3817   6.0   
494    1.0   0.27957   0.0   9.69   0.0  0.585  5.926   42.6  2.3817   6.0   
495    1.0   0.17899   0.0 

In [ ]:
sm.OLS?

In [14]:
model = sm.OLS(y, X_constant)

In [15]:
lr = model.fit()

In [16]:
lr.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.741
Model:                            OLS   Adj. R-squared:                  0.734
Method:                 Least Squares   F-statistic:                     108.1
Date:                Sun, 08 Oct 2017   Prob (F-statistic):          6.95e-135
Time:                        16:51:09   Log-Likelihood:                -1498.8
No. Observations:                 506   AIC:                             3026.
Df Residuals:                     492   BIC:                             3085.
Df Model:                          13                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         36.4911      5.104      7.149      0.000      26.462      46.520
CRIM          -0.1072      0.033     -3.276      0.001      -0.171      -0.043
ZN             0.0464      0.014      3.380      0.001       0.019       0.073
INDUS          0.0209      0.061      0.339      0.735      -0.100       0.142
CHAS           2.6886      0.862      3.120      0.002       0.996       4.381
NOX          -17.7958      3.821     -4.658      0.000     -25.302     -10.289
RM             3.8048      0.418      9.102      0.000       2.983       4.626
AGE            0.0008      0.013      0.057      0.955      -0.025       0.027
DIS           -1.4758      0.199     -7.398      0.000      -1.868      -1.084
RAD            0.3057      0.066      4.608      0.000       0.175       0.436
TAX           -0.0123      0.004     -3.278      0.001      -0.020      -0.005
PTRATIO       -0.9535      0.131     -7.287      0.000      -1.211      -0.696
B              0.0094      0.003      3.500      0.001       0.004       0.015
LSTAT         -0.5255      0.051    -10.366      0.000      -0.625      -0.426
==============================================================================
Omnibus:                      178.029   Durbin-Watson:                   1.078
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              782.015
Skew:                           1.521   Prob(JB):                    1.54e-170
Kurtosis:                       8.276   Cond. No.                     1.51e+04
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.51e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

There are a lot of statistical tests and information. Mostly for the purpose of statistical analysis.

You do not need all of these for data science.

Data science focus is on prediction and having models that work on predicting real data. It is not concerned as much with  correct specifications of statistical problems.

## Model Statistical Outputs:

**Dep. Variable**: The dependent variable or target variable

**Model**: Highlight the model used to obtain this output. It is OLS here. Ordinary least squares / Linear regression

**Method**: The method used to fit the data to the model. Least squares

**No. Observations**: The number of observations

**DF Residuals**: The degrees of freedom of the residuals. Calculated by taking the number of observations less the number of parameters

**DF Model**: The number of estimated parameters in the model. In this case 13. The constant term is not included.


**R-squared**: This is the coefficient of determination. Measure of goodness of fit.
$$R^2=1-\frac{SS_{res}}{SS_{tot}}$$

> From [wiki](https://en.wikipedia.org/wiki/Coefficient_of_determination),

  > The total sum of squares, $SS_{tot}=\sum_i(y_i-\bar{y})^2$

  > The regression sum of squares (explained sum of squares), $SS_{reg}=\sum_i(f_i-\bar{y})^2$

  > The sum of squares of residuals (residual sum of squares), $SS_{res}=\sum_i(y_i-f_i)^2 = \sum_ie^2_i$

**Adj. R-squared**: This is the adjusted R-squared. It is the coefficient of determination adjusted by sample size and the number of parameters used.
$$\bar{R}^2=1-(1-R^2)\frac{n-1}{n-p-1}$$

> $p$ = The total number of explanatory variables not including the constant term

> $n$ = The sample size

**F-statistic**: A measure that tells you if you model is different from a simple average.

**Prob (F-statistic)**: This measures the significance of your F-statistic. Also called p-value of F-statistic. In statistics, p-value equal or lower than 0.05 is considered significant.

**AIC**: This is the Akaike Information Criterion. It evaluatess the model based on the model complexity and number of observations. The lower the better. 

**BIC**: This is the Bayesian Information Criterion. Similar to AIC, except it pushishes models with more parameters.

## Parameters Estimates and the Associated Statistical Tests

**coef**: The estimated coefficient. Note that this is just a point estimate.

**std err**: The standard error of the estimate of the coefficient. Another term for standard deviation

**t**: The t-statistic score. 

**P > |t|**: The p-value. A measure of the probability that the coefficient is different from zero.

**[95.0% Conf. Interval]**: The 95% confidence interval of the coefficient. Shown here as [0.025, 0.975], the lower and upper bound.

## Residual Tests

**Omnibus D'Angostino's test**: This is a combined statistical test for skewness and kurtosis.

**Prob(Omnibus)**: p-value of Omnibus test.

**Skewness**: This is a measure of the symmetry of the residuals around the mean. Zero if symmetrical. A positive value indicates a long tail to the right; a negative value a long tail to the left.

**Kurtosis**: This is a measure of the shape of the distribution of the residuals. A normal distribution has a zero measure. A negative value points to a flatter than normal distribution; a positive one has a higher peak than normal distribution.

**Durbin-Watson**: This is a test for the presence of correlation among the residuals. This is especially important for time series modelling

**Jarque-Bera**: This is a combined statistical test of skewness and kurtosis.

**Prob (JB)**: p-value of Jarque-Bera.

**Cond. No**: This is a test for multicollinearity. > 30 indicates unstable results

***

# statsmodels.formula.api

In [17]:
form_lr = smf.ols(formula = 'y ~ CRIM + ZN + INDUS + CHAS + NOX + RM + AGE + DIS + RAD + TAX + PTRATIO + B + LSTAT', 
              data=df)
mlr = form_lr.fit()

In [18]:
mlr.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.741
Model:                            OLS   Adj. R-squared:                  0.734
Method:                 Least Squares   F-statistic:                     108.1
Date:                Sun, 08 Oct 2017   Prob (F-statistic):          6.95e-135
Time:                        17:00:49   Log-Likelihood:                -1498.8
No. Observations:                 506   AIC:                             3026.
Df Residuals:                     492   BIC:                             3085.
Df Model:                          13                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     36.4911      5.104      7.149      0.000      26.462      46.520
CRIM          -0.1072      0.033     -3.276      0.001      -0.171      -0.043
ZN             0.0464      0.014      3.380      0.001       0.019       0.073
INDUS          0.0209      0.061      0.339      0.735      -0.100       0.142
CHAS           2.6886      0.862      3.120      0.002       0.996       4.381
NOX          -17.7958      3.821     -4.658      0.000     -25.302     -10.289
RM             3.8048      0.418      9.102      0.000       2.983       4.626
AGE            0.0008      0.013      0.057      0.955      -0.025       0.027
DIS           -1.4758      0.199     -7.398      0.000      -1.868      -1.084
RAD            0.3057      0.066      4.608      0.000       0.175       0.436
TAX           -0.0123      0.004     -3.278      0.001      -0.020      -0.005
PTRATIO       -0.9535      0.131     -7.287      0.000      -1.211      -0.696
B              0.0094      0.003      3.500      0.001       0.004       0.015
LSTAT         -0.5255      0.051    -10.366      0.000      -0.625      -0.426
==============================================================================
Omnibus:                      178.029   Durbin-Watson:                   1.078
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              782.015
Skew:                           1.521   Prob(JB):                    1.54e-170
Kurtosis:                       8.276   Cond. No.                     1.51e+04
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.51e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

## Exercise

In [19]:
form_lr = smf.ols(formula = 'y ~ CRIM + ZN + INDUS + CHAS + NOX + RM', 
              data=df)
mlr = form_lr.fit()
mlr.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.587
Model:                            OLS   Adj. R-squared:                  0.582
Method:                 Least Squares   F-statistic:                     118.0
Date:                Sun, 08 Oct 2017   Prob (F-statistic):           2.02e-92
Time:                        17:02:51   Log-Likelihood:                -1616.7
No. Observations:                 506   AIC:                             3247.
Df Residuals:                     499   BIC:                             3277.
Df Model:                           6                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept    -17.8794      3.216     -5.560      0.000     -24.198     -11.561
CRIM          -0.1738      0.035     -5.024      0.000      -0.242      -0.106
ZN             0.0212      0.014      1.526      0.128      -0.006       0.048
INDUS         -0.1444      0.064     -2.257      0.024      -0.270      -0.019
CHAS           4.8009      1.060      4.530      0.000       2.719       6.883
NOX           -7.3163      3.694     -1.981      0.048     -14.574      -0.059
RM             7.3402      0.418     17.577      0.000       6.520       8.161
==============================================================================
Omnibus:                      218.149   Durbin-Watson:                   0.849
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             1523.465
Skew:                           1.732   Prob(JB):                         0.00
Kurtosis:                      10.763   Cond. No.                         420.
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

***

# Correlation Matrix

Useful diagnostic tool to identify collinearity between predictors



In [ ]:
pd.options.display.float_format = '{:,.4f}'.format
corr_matrix = df.corr()
corr_matrix

In [ ]:
corr_matrix[np.abs(corr_matrix) < 0.6] = 0
corr_matrix

In [ ]:
plt.figure(figsize=(16,10))
sns.heatmap(corr_matrix, annot=True, cmap='YlGnBu')
plt.show()

## Detecting Collinearity with Eigenvectors

In [ ]:
eigenvalues, eigenvectors = np.linalg.eig(df.corr())

In [ ]:
pd.Series(eigenvalues).sort_values()

Note that index 8, eigenvalue of 0.0635, is near to zero or very small compared to the others. Small value represents presence of collinearity. 

In [ ]:
np.abs(pd.Series(eigenvectors[:,8])).sort_values(ascending=False)

Note that index 9, 8, 2 have very high loading when compared against the rest

In [ ]:
print(df.columns[2], df.columns[8], df.columns[9])

These are the factors that are causing multicollinearity problem.

***

# Revisiting Feature Importance and Extractions

Check:

1. Direction of the coefficient
2. Impact of the variable / factor on the model

## Standardise Variable to Identify Key Feature(s)

In order to perform point 2 properly, one needs to standardise the variable

In [ ]:
from sklearn.linear_model import LinearRegression

In [ ]:
model = LinearRegression()

In [ ]:
model.fit(X,y)

In [ ]:
result = pd.DataFrame(list(zip(model.coef_, df.columns)), columns=['coefficient', 'name']).set_index('name')
np.abs(result).sort_values(by='coefficient', ascending=False)

In [ ]:
from sklearn.preprocessing import StandardScaler  
from sklearn.pipeline import make_pipeline  
scaler = StandardScaler()  
Stand_coef_linear_reg = make_pipeline(scaler, model)

In [ ]:
Stand_coef_linear_reg.fit(X,y)
result = pd.DataFrame(list(zip(Stand_coef_linear_reg.steps[1][1].coef_, df.columns)), 
                      columns=['coefficient', 'name']).set_index('name')
np.abs(result).sort_values(by='coefficient', ascending=False)

# Use $R^2$ to Identify Key Features

* Compare $R^2$ of model against $R^2$ of model without a feature. 

* A significant change in $R^2$ signify the importance of the feature.




In [ ]:
from sklearn.metrics import r2_score

In [ ]:
linear_reg = smf.ols(formula = 'y ~ CRIM + ZN + INDUS + CHAS + NOX + RM + AGE + DIS + RAD + TAX + PTRATIO + B + LSTAT', 
              data=df)
benchmark = linear_reg.fit()
r2_score(y, benchmark.predict(df))

without LSTAT

In [ ]:
linear_reg = smf.ols(formula = 'y ~ CRIM + ZN + INDUS + CHAS + NOX + RM + AGE + DIS + RAD + TAX + PTRATIO + B', 
              data=df)
lr_without_LSTAT = linear_reg.fit()
r2_score(y, lr_without_LSTAT.predict(df))

without AGE

In [ ]:
linear_reg = smf.ols(formula = 'y ~ CRIM + ZN + INDUS + CHAS + NOX + RM + DIS + RAD + TAX + PTRATIO + B + LSTAT', 
              data=df)
lr_without_AGE = linear_reg.fit()
r2_score(y, lr_without_AGE.predict(df))

***